File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int lcs(string& a, string& b, int n, int m, vector<vector<int>>&dp){
        if(n==0 || m==0) return 0;
        
        if(dp[n][m]!=-1) return dp[n][m];
        
        if(a[n-1] ==b[m-1]){
            return dp[n][m]= lcs(a,b,n-1,m-1,dp)+1;
        }
        else{
            return dp[n][m]= max(lcs(a,b,n-1,m,dp),lcs(a,b,n,m-1,dp));
        }
    }
    int longestCommonSubsequence(string t1, string t2) {
        vector<vector<int>>dp;
        dp.resize(t1.size()+1,vector<int>(t2.size()+1,-1));
        return lcs(t1,t2,t1.size(),t2.size(),dp);
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        @cache
        def dp(a,n1,b,n2):
            if n1==0 or n2==0: return 0
            if a[n1-1]==b[n2-1]: return 1+dp(a,n1-1,b,n2-1)
            else: return max(dp(a,n1,b,n2-1), dp(a,n1-1,b,n2))
        return dp(text1,len(text1),text2,len(text2))